base/heartbeat.go

```diff
@@ -300,7 +287,157 @@ func (h *couchbaseHeartBeater) upsertHeartbeatTimeoutDoc(intervalSeconds int) er
 
 }
 
-func (h *couchbaseHeartBeater) addHeartbeatCheckView() error {
+// documentBackedNodeListHandler tracks nodes in a single node list document
+type documentBackedNodeListHandler struct {
+	nodeListKey string     // key for the tracking document
+	bucket      Bucket     // bucket used for document storage
+	nodeIDs     []string   // Set of nodes from the latest retrieval
+	cas         uint64     // CAS from latest retrieval of tracking document
+	lock        sync.Mutex // lock for nodes access
+}
+
+func NewDocumentBackedNodeListHandler(bucket Bucket, keyPrefix string) (*documentBackedNodeListHandler, error) {
+
+	handler := &documentBackedNodeListHandler{
+		nodeListKey: keyPrefix + "HeartbeatNodeList",
+		bucket:      bucket,
+	}
+	return handler, nil
+}
+
+// Adds the node to the tracking document
+func (dh *documentBackedNodeListHandler) AddNode(nodeID string) error {
+	return dh.updateNodeList(nodeID, false)
+}
+
+// Removes the node to the tracking document
+func (dh *documentBackedNodeListHandler) RemoveNode(nodeID string) error {
+	return dh.updateNodeList(nodeID, true)
+}
+
+func (dh *documentBackedNodeListHandler) GetNodes() ([]string, error) {
+	dh.lock.Lock()
+	err := dh.loadNodeIDs()
+	dh.lock.Unlock()
+	return dh.nodeIDs, err
+}
+
+// Adds or removes a nodeID from the node list document
+func (dh *documentBackedNodeListHandler) updateNodeList(nodeID string, remove bool) error {
+
+	dh.lock.Lock()
+	defer dh.lock.Unlock()
+
+	// Retry loop handles CAS failure
+	for {
+		// Reload the node set if it hasn't been initialized (or has been marked out of date by previous CAS write failure)
+		if dh.cas == 0 {
+			if err := dh.loadNodeIDs(); err != nil {
+				return err
+			}
+		}
+
+		// Check whether nodeID already exists in the set
+		nodeIndex := -1
+		for index, existingNodeID := range dh.nodeIDs {
+			if existingNodeID == nodeID {
+				nodeIndex = index
```

Comment:
break out of loop once found

```suggestion
				nodeIndex = index
				break
```

Reply:
Done